zhmcclient_mock/_hmc.py

```diff
@@ -0,0 +1,1006 @@
+    """
+
+    api_root = '/api'  # root of all resource URIs
+    next_oid = 1  # next object ID, for auto-generating them
+
+    def __init__(self, hmc, parent, resource_class, base_uri, oid_prop,
+                 uri_prop):
+        self._hmc = hmc
+        self._parent = parent
+        self._resource_class = resource_class
+        self._base_uri = base_uri  # Base URI for resources of this type
+        self._oid_prop = oid_prop
+        self._uri_prop = uri_prop
+        self._resources = OrderedDict()  # Resource objects, by object ID
+
+    @property
+    def hmc(self):
+        """
+        The faked HMC this manager is part of (an object of
+        :class:`~zhmcclient_mock.FakedHmc`).
+        """
+        return self._hmc
+
+    @property
+    def parent(self):
+        """
+        The parent (scoping resource) for this manager (an object of a derived
+        class of :class:`~zhmcclient_mock.FakedBaseResource`).
+        """
+        return self._parent
+
+    @property
+    def resource_class(self):
+        """
+        The resource class managed by this manager (a derived class of
+        :class:`~zhmcclient_mock.FakedBaseResource`).
+        """
+        return self._resource_class
+
+    @property
+    def base_uri(self):
+        """
+        The base URI for URIs of resources managed by this manager.
+        """
+        return self._base_uri
+
+    @property
+    def oid_prop(self):
+        """
+        The name of the resource property for the object ID ('object-id' or
+        'element-id').
+        """
+        return self._oid_prop
+
+    @property
+    def uri_prop(self):
+        """
+        The name of the resource property for the object URI ('object-uri' or
+        'element-uri').
+        """
+        return self._uri_prop
+
+    def _new_oid(self):
+        new_oid = self.next_oid
+        self.next_oid += 1
+        return str(new_oid)
+
+    def add(self, properties):
+        """
+        Add a faked resource to this manager.
+
+        For URI-based lookup, the resource is also added to the faked HMC.
+
+        Parameters:
+
+          properties (dict):
+            Resource properties. If the URI property (e.g. 'object-uri') or the
+            object ID property (e.g. 'object-id') are not specified, they
+            will be auto-generated.
+
+        Returns:
+          FakedBaseResource: The faked resource object.
+        """
+        resource = self.resource_class(self, properties)
+        self._resources[resource.oid] = resource
+        self._hmc._resources[resource.uri] = resource
+        return resource
+
+    def remove(self, oid):
+        """
+        Remove a faked resource from this manager.
+
+        Parameters:
+
+          oid (string):
+            The object ID of the resource (e.g. value of the 'object-uri'
+            property).
+        """
+        uri = self._resources[oid].uri
+        del self._resources[oid]
+        del self._hmc._resources[uri]
+
```

Comment:
Preethi, `find()` and `findall()` are supported.
The mock layer works such that its `FakedSession` is instantiated by the testcase using the mock layer instead of the real `Session`. For example, the find() function on CPCs is still available in the same way it would be with a real `Session`, but when going through the `FakedSession`, it becomes a `list()` on the `FakedCpcManager` of the `FakedHmc`. The bottom line is, the methods of the `FakedHmc` class are low level add/remove/list methods that are used internally by the mock layer, and therefore it is not meant to be used directly and therefore not compatible with the external API.